### PR TITLE
fix(acir_gen): Pass accurate contents to slice inputs for bb func calls

### LINKED ISSUE #2421: Using the resulting slice from to_radix in signature verification fails the verification
### Aim

Tried to use the resulting bytes from a field decomposition as input to ECDSA & schnorr signature verification.

### Expected Behavior

The verification should pass the same as if the decomposition was done outside the circuit.

### Bug

Try executing this modified schnorr e2e test:
```rust
use dep::std;

// Note: If main has any unsized types, then the verifier will never be able
// to figure out the circuit instance
fn main(message: [u8; 10], message_field: Field, pub_key_x: Field, pub_key_y: Field, signature: [u8; 64]) {
    let message_field_bytes = message_field.to_be_bytes(10);
    for i in 0..10 {
        assert(message[i] == message_field_bytes[i]);
    }
    // Is there ever a situation where someone would want 
    // to ensure that a signature was invalid?
    let valid_signature = std::schnorr::verify_signature(pub_key_x,pub_key_y,signature, message_field_bytes);
    assert(valid_signature);
}
```
If you try to pass message_field_bytes it will fail, but if you pass message it will pass.
Prover.toml:
```toml
message = [0,1,2,3,4,5,6,7,8,9]
message_field = "0x010203040506070809"
pub_key_x = "0x17cbd3ed3151ccfd170efe1d54280a6a4822640bf5c369908ad74ea21518a9c5"
pub_key_y = "0x0e0456e3795c1a31f20035b741cd6158929eeccd320d299cfcac962865a6bc74"
signature = [
    5, 202, 31, 146, 81, 242, 246, 69, 43, 107, 249, 153, 198, 44, 14, 111, 191, 121, 137, 166, 
    160, 103, 18, 181, 243, 233, 226, 95, 67, 16, 37, 128, 85, 76, 19, 253, 30, 77, 192, 53, 138, 
    205, 69, 33, 236, 163, 83, 194, 84, 137, 184, 221, 176, 121, 179, 27, 63, 70, 54, 16, 176, 
    250, 39, 239,
]
```

This is a regression, since using the output of to_be_bytes as input to a signature verification was working previously in aztec account contracts. This started happening after the rework of slices https://github.com/noir-lang/noir/pull/2347 but looking at the PR content I don't understand why signature verifications would fail using the new slices :thinking: 

### To Reproduce

1.
2.
3.
4.

### Installation Method

Compiled from source

### Nargo Version

_No response_

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by vezenovm**
# Description

## Problem\*

Resolves #2421

## Summary\*

The internal structure of slices in SSA was changed in https://github.com/noir-lang/noir/pull/2347. Slices and arrays are polymorphic over each other, thus slices can be passed to functions that accept arrays with a generic size (ex. fn foo<N>(input: [Field; N]). The slice structure is handled in SSA, but in ACIR gen we setup the inputs to for the black box calls manually. As the slice length always precedes the slice contents, without changes to `call_black_box` in acir gen, the slice length was being passed to the bb funcs rather than the contents of the slice.

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by vezenovm** [comment]
@sirasistant I added the test from the issue, but it would be great if you could test this fix against the sha256 + ecdsa as well

**Comment by sirasistant** [comment]
Will test it! I think this might also be an issue in brillig? 🤔

**Comment by vezenovm** [comment]
> Will test it! I think this might also be an issue in brillig?

Yeah it is, I am fixing

**Comment by vezenovm** [comment]
@sirasistant I have updated for brillig as well now. This should be ready for review

**Comment by sirasistant** [comment]
It's fixed on aztec packages! :partying_face: reviewing now :)

File: crates/noirc_evaluator/src/brillig/brillig_gen/brillig_black_box.rs [hunk 1]
```diff
@@ -16,12 +16,24 @@ pub(crate) fn convert_black_box_call(
...
-                [RegisterOrMemory::HeapArray(result_array)],
-            ) = (function_arguments, function_results)
+            if let ([..], [RegisterOrMemory::HeapArray(result_array)]) =
+                (function_arguments, function_results)
             {
-                let message_vector = brillig_context.array_to_vector(message_array);
+                // Slices are represented as a tuple of (length, slice contents).
+                // We must check the number of inputs to differentiate between arrays and slices
+                // and make sure that we pass the correct inputs to the function call.
+                let message = if function_arguments.len() > 1 {
```
Comment (sirasistant) [inline crates/noirc_evaluator/src/brillig/brillig_gen/brillig_black_box.rs:25]: Can't help but think that we should engineer a solution for reading arguments that makes this easier to handle :thinking:

**Review by sirasistant — APPROVED** [review]
Thanks for the fix! We might want in the future to abstract this a bit, especially if we end up having a blackbox function that accepts more than one slice, It'd be complex to handle this way

File: crates/nargo_cli/tests/execution_success/schnorr/src/main.nr [hunk 2]
```diff
@@ -2,9 +2,15 @@ use dep::std;
...
+    // We want to make sure that we can accurately verify a signature whose message is a slice vs. an array
+    let message_field_bytes = message_field.to_be_bytes(10);
+    for i in 0..10 {
+        assert(message[i] == message_field_bytes[i]);
+    }
     // Is there ever a situation where someone would want 
     // to ensure that a signature was invalid?
-    let valid_signature = std::schnorr::verify_signature(pub_key_x,pub_key_y,signature, message);
+    let valid_signature = std::schnorr::verify_signature(pub_key_x,pub_key_y,signature, message_field_bytes);
     assert(valid_signature);
```
[the remark this answers is not in the document]
Comment (vezenovm) [reply crates/nargo_cli/tests/execution_success/schnorr/src/main.nr:15]: Sure I will add that quick

**Comment by vezenovm** [comment]
> We might want in the future to abstract this a bit, especially if we end up having a blackbox function that accepts more than one slice, It'd be complex to handle this way

I agree. I more meant for this to be a quick solution as I work towards other priorities. I will make an issue to improve the handling of black box inputs

**Comment by vezenovm** [comment]
@sirasistant I just pushed an update which dismissed your review, could you take another look?

**Review by sirasistant — APPROVED** [review]

**Pull request merged**